File: app/src/main/java/com/musicplayer/ui/adapter/SongAdapter.java
package com.musicplayer.ui.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.bumptech.glide.Glide;
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions;
import com.musicplayer.R;
import com.musicplayer.model.Song;
import java.util.ArrayList;
import java.util.List;

public class SongAdapter extends RecyclerView.Adapter<SongAdapter.SongViewHolder> {
    private List<Song> songs = new ArrayList<>();
    private final OnSongActionListener listener;
    private boolean playlistMode = false;

    public interface OnSongActionListener {
        void onPlay(Song song);
        void onDownload(Song song);
        void onDelete(Song song);
        void onAddToPlaylist(Song song);
    }

    public SongAdapter(OnSongActionListener listener) {
        this.listener = listener;
    }

    @NonNull
    @Override
    public SongViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
            .inflate(R.layout.item_song, parent, false);
        return new SongViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull SongViewHolder holder, int position) {
        Song song = songs.get(position);
        holder.tvTitle.setText(song.getTitle());
        holder.tvArtist.setText(song.getArtist());
        
        if (song.isLocal()) {
            holder.btnDownload.setImageResource(android.R.drawable.ic_menu_delete);
            holder.btnDownload.setVisibility(View.VISIBLE);
            holder.btnDownload.setOnClickListener(v -> listener.onDelete(song));
        } else {
            holder.btnDownload.setImageResource(android.R.drawable.stat_sys_download_done); // O el que tuviera antes
            // Nota: El XML original probablemente tiene el icono de descarga.
            // Vamos a usar uno estándar de descarga si no es local.
            holder.btnDownload.setImageResource(android.R.drawable.stat_sys_download);
            holder.btnDownload.setVisibility(View.VISIBLE);
            holder.btnDownload.setOnClickListener(v -> listener.onDownload(song));
        }
        
        Glide.with(holder.itemView.getContext())
            .load(song.getThumbnailUrl())
            .placeholder(android.R.drawable.ic_menu_report_image)
            .transition(DrawableTransitionOptions.withCrossFade())
            .centerCrop()
            .into(holder.imgThumbnail);
            
        holder.itemView.setOnClickListener(v -> listener.onPlay(song));
        holder.btnMore.setOnClickListener(v -> {
            androidx.appcompat.widget.PopupMenu popup = new androidx.appcompat.widget.PopupMenu(v.getContext(), v);
            popup.getMenu().add("Añadir a Playlist");
            popup.getMenu().add("Descargar");
            if (playlistMode) {
                popup.getMenu().add("Eliminar de esta Playlist");
            }
            
            popup.setOnMenuItemClickListener(item -> {
                String title = item.getTitle().toString();
                if (title.equals("Añadir a Playlist")) {
                    listener.onAddToPlaylist(song);
                } else if (title.equals("Descargar")) {
                    listener.onDownload(song);
                } else if (title.equals("Eliminar de esta Playlist")) {
                    listener.onDelete(song);
                }
                return true;
            });
            popup.show();
        });
        
        // Ya no necesitamos botones de descarga directos si usamos el menú
        holder.btnDownload.setVisibility(View.GONE);
    }

    @Override
    public int getItemCount() {
        return songs.size();
    }

    public void updateSongs(List<Song> newSongs) {
        songs.clear();
        songs.addAll(newSongs);
        notifyDataSetChanged();
    }

    public List<Song> getSongs() {
        return songs;
    }

    public void setPlaylistMode(boolean playlistMode) {
        this.playlistMode = playlistMode;
    }

    static class SongViewHolder extends RecyclerView.ViewHolder {
        ImageView imgThumbnail;
        TextView tvTitle;
        TextView tvArtist;
        ImageButton btnDownload;
        ImageButton btnMore;

        public SongViewHolder(@NonNull View itemView) {
            super(itemView);
            imgThumbnail = itemView.findViewById(R.id.imgSongThumbnail);
            tvTitle = itemView.findViewById(R.id.tvSongTitle);
            tvArtist = itemView.findViewById(R.id.tvSongArtist);
            btnDownload = itemView.findViewById(R.id.btnDownloadItem);
            btnMore = itemView.findViewById(R.id.btnMoreItem);
        }
    }
}

File: app/src/main/java/com/musicplayer/ui/adapter/SongAdapter.kt
package com.musicplayer.ui.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
import com.musicplayer.R
import com.musicplayer.model.Song
import java.util.*

class SongAdapter(private val listener: OnSongActionListener) :
    ListAdapter<Song, RecyclerView.ViewHolder>(SongDiffCallback()) {

    private var playlistMode = false
    private var selectionMode = false
    private val selectedSongs = mutableSetOf<Song>()
    var onSelectionChanged: ((Int) -> Unit)? = null

    fun setSelectionMode(enabled: Boolean) {
        selectionMode = enabled
        if (!enabled) selectedSongs.clear()
        notifyDataSetChanged()
    }

    fun getSelectedSongs(): List<Song> = selectedSongs.toList()

    interface OnSongActionListener {
        fun onPlay(song: Song)
        fun onDownload(song: Song)
        fun onDelete(song: Song)
        fun onAddToPlaylist(song: Song)
    }

    companion object {
        private const val TYPE_SONG = 0
        private const val TYPE_HEADER = 1
        private const val TYPE_MASTER = 2
    }

    init {
        setHasStableIds(true)
        stateRestorationPolicy = StateRestorationPolicy.PREVENT_WHEN_EMPTY
    }

    override fun getItemId(position: Int): Long {
        val s = getItem(position) ?: return RecyclerView.NO_ID
        return s.id.hashCode().toLong()
    }

    override fun getItemViewType(position: Int): Int {
        val song = getItem(position)
        return when {
            song.isHeader && selectionMode -> TYPE_MASTER
            song.isHeader -> TYPE_HEADER
            else -> TYPE_SONG
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            TYPE_HEADER -> {
                val view = LayoutInflater.from(parent.context).inflate(R.layout.item_header, parent, false)
                HeaderViewHolder(view)
            }
            TYPE_MASTER -> {
                val view = LayoutInflater.from(parent.context).inflate(R.layout.item_song_master, parent, false)
                SongViewHolder(view)
            }
            else -> {
                val view = LayoutInflater.from(parent.context).inflate(R.layout.item_song, parent, false)
                SongViewHolder(view)
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val song = getItem(position)
        if (holder is HeaderViewHolder) {
            holder.tvTitle.text = song.title
            return
        }

        val songHolder = holder as SongViewHolder
        songHolder.tvTitle.text = song.title
        songHolder.tvArtist.text = song.artist

        val originalThumb = song.thumbnailUrl
        val maxRes = originalThumb?.replace("hqdefault.jpg", "maxresdefault.jpg") ?: originalThumb
        val sdRes = originalThumb?.replace("hqdefault.jpg", "sddefault.jpg") ?: originalThumb

        Glide.with(holder.itemView.context)
            .load(maxRes)
            .error(Glide.with(holder.itemView.context).load(sdRes)
                .error(Glide.with(holder.itemView.context).load(originalThumb)))
            .placeholder(R.drawable.circle_background)
            .diskCacheStrategy(DiskCacheStrategy.RESOURCE)
            .transition(DrawableTransitionOptions.withCrossFade(150))
            .override(200, 200)
            .centerCrop()
            .into(songHolder.imgThumbnail)

        songHolder.checkBox?.let { cb ->
            cb.visibility = if (selectionMode) View.VISIBLE else View.GONE
            cb.setOnCheckedChangeListener(null)
            cb.isChecked = selectedSongs.any { it.id == song.id }
            cb.setOnCheckedChangeListener { _, isChecked ->
                if (isChecked) selectedSongs.add(song) else selectedSongs.remove(song)
                onSelectionChanged?.invoke(selectedSongs.size)
            }
        }

        holder.itemView.setOnClickListener { 
            if (selectionMode) {
                songHolder.checkBox?.toggle()
            } else {
                listener.onPlay(song) 
            }
        }
        holder.itemView.setOnLongClickListener {
            listener.onDelete(song)
            true
        }
        
        songHolder.btnMore?.setOnClickListener { v ->
            val popup = androidx.appcompat.widget.PopupMenu(v.context, v)
            popup.menu.add("Añadir a Playlist")
            popup.menu.add("Descargar")
            if (playlistMode) {
                popup.menu.add("Eliminar de esta Playlist")
            }

            popup.setOnMenuItemClickListener { item ->
                when (item.title.toString()) {
                    "Añadir a Playlist" -> listener.onAddToPlaylist(song)
                    "Descargar" -> listener.onDownload(song)
                    "Eliminar de esta Playlist" -> listener.onDelete(song)
                }
                true
            }
            popup.show()
        }

        songHolder.btnDownload?.visibility = View.GONE
    }

    fun updateSongs(newSongs: List<Song>) {
        submitList(newSongs.toList())
    }

    val songs: List<Song>
        get() = currentList

    fun setPlaylistMode(playlistMode: Boolean) {
        this.playlistMode = playlistMode
    }

    class SongViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val imgThumbnail: ImageView = itemView.findViewById(R.id.imgSongThumbnail)
        val tvTitle: TextView = itemView.findViewById(R.id.tvSongTitle)
        val tvArtist: TextView = itemView.findViewById(R.id.tvSongArtist)
        val btnDownload: ImageButton? = itemView.findViewById(R.id.btnDownloadItem)
        val btnMore: ImageButton? = itemView.findViewById(R.id.btnMoreItem)
        val checkBox: android.widget.CheckBox? = itemView.findViewById(R.id.cbSelect)
    }

    class HeaderViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val tvTitle: TextView = itemView.findViewById(R.id.tvHeaderTitle)
    }

    private class SongDiffCallback : DiffUtil.ItemCallback<Song>() {
        override fun areItemsTheSame(oldItem: Song, newItem: Song): Boolean {
            return oldItem.id == newItem.id
        }

        override fun areContentsTheSame(oldItem: Song, newItem: Song): Boolean {
            return oldItem.title == newItem.title &&
                    oldItem.artist == newItem.artist &&
                    oldItem.thumbnailUrl == newItem.thumbnailUrl
        }
    }
}
